### PR TITLE
Add BYTE_STREAM_SPLIT support to Parquet

### LINKED ISSUE #15226: [FEA] Parquet reader and writer should support BYTE_STREAM_SPLIT encoding
**Is your feature request related to a problem? Please describe.**
BYTE_STREAM_SPLIT encoding is the only Parquet encoding left that cuDF does not support. Previously limited to use with FLOAT and DOUBLE columns, there is a current proposal (https://github.com/apache/parquet-format/pull/229, and [jira](https://issues.apache.org/jira/browse/PARQUET-2414)) to extend this encoding to all fixed-width data types. When coupled with compression this encoding can provide significant space savings, and its implementation is less CPU intensive than DELTA_BINARY_PACKED encoding.

**Describe the solution you'd like**
Implement BYTE_STREAM_SPLIT encoding and decoding in cuDF.

**Describe alternatives you've considered**
Not supporting this encoding.

**Additional context**
This encoding is a pretty straightforward reimplementation of PLAIN encoding, but it may lead to memory issues due to the need to access input/output buffers with a large stride.

### PR DESCRIPTION
**Pull request by etseidl**
## Description
Closes #15226. Part of #13501.  Adds support for reading and writing `BYTE_STREAM_SPLIT` encoded Parquet data. Includes a "microkernel" version like those introduced by #15159.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by etseidl** [comment]
`BYTE_STREAM_SPLIT` was originally just for floating point, but it is being expanded to all fixed-width types (https://github.com/apache/parquet-format/pull/229). I did a quick comparison with some mixed double/integer data to demonstrate the utility of the new encoding. While the uncompressed size of the data matches `PLAIN` encoding, when coupled with compression `BYTE_STREAM_SPLIT` can outperform the other encoding options in terms of both speed and compressed data size. The one exception being uncompressed `DELTA_BINARY_PACKED` which achieves a data reduction comparable to ZSTD compression, for a far lower computational cost.

```
             dict   plain   delta    bss   delta (no comp)
decompress   214ms     52     217     26      0
decode         8ms      6      10      9     10
encode        13ms      6      13      7     17
compress     438ms    369     333    223      0
round trp   1155ms    800     972    611    465
uncomp sz     25MB     40      18     40     18
comp sz       16MB     15      16     14     18
```

**Comment by etseidl** [comment]
Implementation note: I went with separate encode and decode kernels, rather than modifying the PLAIN kernels. My thinking was there is already branching to support PLAIN and DICTIONARY decoding, and adding a third option would make already complicated kernels even more so. Easy enough to switch if that fear is unfounded.

File: cpp/src/io/parquet/page_data.cuh [hunk 1]
```diff
@@ -397,4 +397,80 @@ inline __device__ void gpuOutputGeneric(
...
+  T unscaled = 0;
+  // fixed_len_byte_array decimals are big endian
+  for (unsigned int i = 0; i < dtype_len_in; i++) {
+    unscaled = (unscaled << 8) | src[i * stride];
+  }
+  // Shift the unscaled value up and back down when it isn't all 8 bytes,
+  // which sign extend the value for correctly representing negative numbers.
+  if (dtype_len_in < sizeof(T)) {
+    unscaled <<= (sizeof(T) - dtype_len_in) * 8;
+    unscaled >>= (sizeof(T) - dtype_len_in) * 8;
```
Comment (vuule) [inline cpp/src/io/parquet/page_data.cuh:471]: lol

**Review by vuule — COMMENTED** [review]
partial review
There's more to this than I expected, based on the encode description at https://parquet.apache.org/docs/file-format/data-pages/encodings/#byte-stream-split-byte_stream_split--9

File: cpp/src/io/parquet/writer_impl.cu [hunk 2]
```diff
@@ -611,8 +611,7 @@ std::vector<schema_tree_node> construct_schema_tree(
                                                 column_in_metadata const& col_meta) {
         s.requested_encoding = column_encoding::USE_DEFAULT;
 
-        if (schema[parent_idx].name != "list" and
-            col_meta.get_encoding() != column_encoding::USE_DEFAULT) {
+        if (s.name != "list" and col_meta.get_encoding() != column_encoding::USE_DEFAULT) {
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl.cu:614]: I think this was a bug that wasn't caught before.

File: cpp/tests/io/parquet_writer_test.cpp [hunk 3]
```diff
@@ -1864,6 +1977,34 @@ TYPED_TEST(ParquetWriterTimestampTypeTest, TimestampOverflow)
   CUDF_TEST_EXPECT_TABLES_EQUAL(expected, result.tbl->view());
 }
 
+TYPED_TEST(ParquetWriterTimestampTypeTest, TimestampsByteStreamSplit)
+{
+  auto sequence = cudf::detail::make_counting_transform_iterator(
+    0, [](auto i) { return ((std::rand() / 10000) * 1000); });
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/tests/io/parquet_writer_test.cpp:1983]: Ok. This was just cut-and-pasted from another test.

File: cpp/src/io/parquet/page_data.cuh (hunk 1 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_data.cuh:471]: yeah, seems like a no-op :sweat_smile:

File: cpp/src/io/parquet/parquet_gpu.hpp [hunk 4]
```diff
@@ -753,6 +758,28 @@ void DecodePageData(cudf::detail::hostdevice_span<PageInfo> pages,
...
+ * @param[in,out] pages All pages to be decoded
+ * @param[in] chunks All chunks to be decoded
+ * @param[in] num_rows Total number of rows to read
+ * @param[in] min_row Minimum number of rows to read
+ * @param[in] level_type_size Size in bytes of the type for level decoding
+ * @param[out] error_code Error code for kernel failures
+ * @param[in] stream CUDA stream to use
+ */
+void DecodeSplitPageData(cudf::detail::hostdevice_span<PageInfo>& pages,
+                         cudf::detail::hostdevice_span<ColumnChunkDesc> const& chunks,
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/parquet_gpu.hpp]: Not sure where that came from. Fixed.

File: cpp/src/io/parquet/writer_impl.cu (hunk 2 above)
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl.cu:614]: I don't like where this is heading... :sweat_smile: 

Let me check. I don't remember if a current test caught it or a local test file.

[the remark this answers is not in the document]
Comment (vuule) [reply cpp/src/io/parquet/writer_impl.cu:614]: :joy:
I knew it! Trying to sneak bug fixes past me, again?
five extra nitpicks for you

[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/writer_impl.cu:614]: :sob: 
Added to an existing test.

File: cpp/src/io/parquet/page_enc.cu [hunk 5]
```diff
@@ -1738,16 +1762,29 @@ CUDF_KERNEL void __launch_bounds__(block_size, 8)
...
+          } else {
+            dst[pos + 0] = v;
+            dst[pos + 1] = v >> 8;
+            dst[pos + 2] = v >> 16;
+            dst[pos + 3] = v >> 24;
+            dst[pos + 4] = v >> 32;
+            dst[pos + 5] = v >> 40;
+            dst[pos + 6] = v >> 48;
+            dst[pos + 7] = v >> 56;
+          }
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:1784]: unless we're really worried about the extra multiplication, we can avoid the extra repetition here
```suggestion
        auto const stride     = is_split_stream ? s->page.num_valid : 1;
        dst[pos + 0 * stride] = v;
        dst[pos + 1 * stride] = v >> 8;
        dst[pos + 2 * stride] = v >> 16;
        dst[pos + 3 * stride] = v >> 24;
        dst[pos + 4 * stride] = v >> 32;
        dst[pos + 5 * stride] = v >> 40;
        dst[pos + 6 * stride] = v >> 48;
        dst[pos + 7 * stride] = v >> 56;
```

File: cpp/src/io/parquet/page_enc.cu [hunk 6]
```diff
@@ -1791,10 +1828,24 @@ CUDF_KERNEL void __launch_bounds__(block_size, 8)
...
+            dst[pos + 1 * stride] = v >> 8;
+            dst[pos + 2 * stride] = v >> 16;
+            dst[pos + 3 * stride] = v >> 24;
+            dst[pos + 4 * stride] = v >> 32;
+            dst[pos + 5 * stride] = v >> 40;
+            dst[pos + 6 * stride] = v >> 48;
+            dst[pos + 7 * stride] = v >> 56;
+          } else {
+            auto v = s->col.leaf_column->element<double>(val_idx);
+            memcpy(dst + pos, &v, 8);
```
Comment (vuule) [inline cpp/src/io/parquet/page_enc.cu:1844]: wait, is this the same as the numerous `dst[pos + x] = v >> x*8;` ?

File: cpp/src/io/parquet/page_enc.cu (hunk 5 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu:1784]: I'll benchmark it. Hopefully the compiler is smart enough to figure out it can skip the multiplication by 1.

Comment (vuule) [reply cpp/src/io/parquet/page_enc.cu:1784]: there's also MAD :crossed_fingers:

**Review by vuule — COMMENTED** [review]
rest of the review; just a few typos, looks real solid

File: cpp/src/io/parquet/decode_fixed.cu [hunk 7]
```diff
@@ -495,6 +585,123 @@ CUDF_KERNEL void __launch_bounds__(decode_block_size)
...
+  gpuDecodeSplitPageDataFlat(PageInfo* pages,
+                             device_span<ColumnChunkDesc const> chunks,
+                             size_t min_row,
+                             size_t num_rows,
+                             kernel_error::pointer error_code)
+{
+  __shared__ __align__(16) page_state_s state_g;
+  __shared__ __align__(16) page_state_buffers_s<rolling_buf_size,  // size of nz_idx buffer
+                                                1,                 // unused in this kernel
+                                                1>                 // unused in this kernel
```
Comment (vuule) [inline cpp/src/io/parquet/decode_fixed.cu:613]: seems like these can be zeroes, same with line 485

Comment (etseidl) [reply cpp/src/io/parquet/decode_fixed.cu:613]: The template params are used to size an array, so I'm not sure if it is allowed to be 0.

Comment (vuule) [reply cpp/src/io/parquet/decode_fixed.cu:613]: I wasn't sure either, but it seems to work (on my system :smiling_imp:)
Also, some other kernels use zeros for these, e.g. `gpuDecodeDeltaBinary`

File: cpp/src/io/parquet/page_data.cu [hunk 8]
```diff
@@ -28,6 +28,177 @@ namespace {
...
+      int dst_pos = sb->nz_idx[rolling_index<rolling_buf_size>(src_pos)];
+
+      // for the flat hierarchy case we will be reading from the beginning of the value stream,
+      // regardless of the value of first_row. so adjust our destination offset accordingly.
+      // example:
+      // - user has passed skip_rows = 2, so our first_row to output is 2
+      // - the row values we get from nz_idx will be
+      //   0, 1, 2, 3, 4 ....
+      // - by shifting these values by first_row, the sequence becomes
+      //   -1, -2, 0, 1, 2 ...
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/page_data.cu:125]: and copied twice at that :)

File: cpp/src/io/parquet/page_enc.cu (hunk 5 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu]: Use a loop for this was a tiny bit faster for one file I've been testing with, and with nvbench there was no real difference observed switching to a loop. I simplified a good bit of the encoding.

File: cpp/src/io/parquet/page_enc.cu (hunk 6 above)
Comment (etseidl) [reply cpp/src/io/parquet/page_enc.cu]: Yes. I've consolidated this with INT64.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 7 above)
Comment (etseidl) [reply cpp/src/io/parquet/decode_fixed.cu:613]: Hmm...I guess I'll try 0s then 😅

Comment (vuule) [reply cpp/src/io/parquet/decode_fixed.cu]: yeah, I'm also surprised it works

Comment (etseidl) [reply cpp/src/io/parquet/decode_fixed.cu]: After some offline discussion/research, we've determined `1` is the correct param for unused arrays. C++ does not support 0 length arrays in most instances, but a compiler extension was allowing it here.

File: cpp/src/io/parquet/decode_fixed.cu [hunk 9]
```diff
@@ -225,6 +225,98 @@ __device__ inline void gpuDecodeValues(
...
+        if (s->dtype_len_in == 4) {
+          // Reading INT32 TIME_MILLIS into 64-bit DURATION_MILLISECONDS
+          // TIME_MILLIS is the only duration type stored as int32:
+          // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype
+          gpuOutputByteStreamSplit<int32_t>(dst, src, num_values);
+          dst[4] = 0;
+          dst[5] = 0;
+          dst[6] = 0;
+          dst[7] = 0;
+        } else if (s->ts_scale) {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/decode_fixed.cu:303]: I may be wrong but wouldn't `ts_scale` also be applicable to `TIME_MILLIS` case when seconds or days are encoded into it such as: https://github.com/mhaseeb123/cudf/blob/branch-24.06/cpp/src/io/parquet/writer_impl.cu#L409

File: cpp/tests/io/parquet_writer_test.cpp [hunk 10]
```diff
@@ -1681,6 +1699,119 @@ TEST_F(ParquetWriterTest, DeltaBinaryStartsWithNulls)
...
+  constexpr cudf::size_type num_rows = 100;
+  auto seq_col0                      = random_values<int32_t>(num_rows);
+  auto seq_col1                      = random_values<int64_t>(num_rows);
+
+  auto durations_ms = cudf::test::fixed_width_column_wrapper<cudf::duration_ms, int64_t>(
+    seq_col0.begin(), seq_col0.end(), no_nulls());
+  auto durations_us = cudf::test::fixed_width_column_wrapper<cudf::duration_us, int64_t>(
+    seq_col1.begin(), seq_col1.end(), no_nulls());
+
+  auto expected = table_view{{durations_ms, durations_us}};
```
Comment (mhaseeb123) [inline cpp/tests/io/parquet_writer_test.cpp:1794]: should we also test `duration_D`, `duration_s` and `duration_ns` for `ts_scale` setting?

File: cpp/src/io/parquet/page_data.cu [hunk 11]
```diff
@@ -28,6 +28,179 @@ namespace {
...
+          if (s->dtype_len_in == 4) {
+            // Reading INT32 TIME_MILLIS into 64-bit DURATION_MILLISECONDS
+            // TIME_MILLIS is the only duration type stored as int32:
+            // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype
+            gpuOutputByteStreamSplit<int32_t>(dst, src, num_values);
+            dst[4] = 0;
+            dst[5] = 0;
+            dst[6] = 0;
+            dst[7] = 0;
+          } else if (s->ts_scale) {
```
Comment (mhaseeb123) [inline cpp/src/io/parquet/page_data.cu:184]: Same question here about the use of `ts_scale` in TIME_MILLIS case as well.

**Review by mhaseeb123 — COMMENTED** [review]
A few comments about the use of `ts_scale` in TIME_MILLIS type as seconds and days are encoded as millis in Parquet plus corresponding time units in tests.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 9 above)
Comment (etseidl) [reply cpp/src/io/parquet/decode_fixed.cu:303]: > Wouldn't ts_scale also be applicable to TIME_MILLIS case when seconds or days are encoded into it

This is on the decode path, and parquet does not have a concept of seconds or days. The `ts_scale` here is derived from the `timestamp_type` parameter in the reader options, and is only valid for timestamps. We'd either have to add another `time_type` option (non-breaking) or re-use `timestamp_type` for `time` fields (breaking) if we wanted to do a similar translation on read.

This is all part of the broader issue of how cuDF handles time types that aren't representable in Parquet (as I know you're aware :wink:).

File: cpp/tests/io/parquet_writer_test.cpp (hunk 10 above)
Comment (etseidl) [reply cpp/tests/io/parquet_writer_test.cpp:1794]: I'm torn...the original point was just to test INT32 vs INT64 handling, especially to exercise the int32->int64 logic when reading the TIME_MILLIS column. Adding `duration_ns` would be easy, but the non-supported types can't just use `CUDF_TEST_EXPECT_TABLES_EQUAL`, but instead have to cast the columns after reading them back in. 

Perhaps instead we could modify `test_durations()` to also use the new encoding in addition to the defaults.

File: cpp/src/io/parquet/decode_fixed.cu (hunk 9 above)
Comment (mhaseeb123) [reply cpp/src/io/parquet/decode_fixed.cu:303]: Ah I see. Thanks for the clarification! 😊

**Comment by mhaseeb123** [comment]
Looks good to me!

**Review by mhaseeb123 — APPROVED** [review]

File: cpp/src/io/parquet/page_data.cu [hunk 12]
```diff
@@ -28,6 +28,179 @@ namespace {
...
+        } else if (dtype_len == 8) {
+          if (s->dtype_len_in == 4) {
+            // Reading INT32 TIME_MILLIS into 64-bit DURATION_MILLISECONDS
+            // TIME_MILLIS is the only duration type stored as int32:
+            // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype
+            gpuOutputByteStreamSplit<int32_t>(dst, src, num_values);
+            dst[4] = 0;
+            dst[5] = 0;
+            dst[6] = 0;
+            dst[7] = 0;
```
Comment (vuule) [inline cpp/src/io/parquet/page_data.cu:183]: memset?

File: cpp/src/io/parquet/decode_fixed.cu [hunk 13]
```diff
@@ -225,6 +225,98 @@ __device__ inline void gpuDecodeValues(
...
+      } else if (dtype_len == 8) {
+        if (s->dtype_len_in == 4) {
+          // Reading INT32 TIME_MILLIS into 64-bit DURATION_MILLISECONDS
+          // TIME_MILLIS is the only duration type stored as int32:
+          // https://github.com/apache/parquet-format/blob/master/LogicalTypes.md#deprecated-time-convertedtype
+          gpuOutputByteStreamSplit<int32_t>(dst, src, num_values);
+          dst[4] = 0;
+          dst[5] = 0;
+          dst[6] = 0;
+          dst[7] = 0;
```
Comment (vuule) [inline cpp/src/io/parquet/decode_fixed.cu:302]: memset?

**Review by vuule — COMMENTED** [review]
looks great, just a few small comments

File: cpp/src/io/parquet/page_data.cuh [hunk 14]
```diff
@@ -396,4 +396,80 @@ inline __device__ void gpuOutputGeneric(
     }
   }
 }
+
+/**
+ * Output a BYTE_STREAM_SPLIT value of length `byte_length`.
+ *
+ * Data is encoded as N streams of length M, forming an NxM sized matrix. Rows are streams,
```
[the remark this answers is not in the document]
Comment (etseidl) [reply cpp/src/io/parquet/page_data.cuh:403]: Yes. Never changed the docstring after I changed the template parameter.

**Review by vuule — APPROVED** [review]
:fire:

**Pull request merged**